### PR TITLE
Fix crash in Muon TF Asummetry

### LINKED ISSUE #23031: Mantid crash fitting Muon data
**Original reporter:** Thomas Jubb

### Expected behavior

Mantid should not crash - it does not crash in the previous version of Mantid.

### Actual behavior

Mantid hard crashes

### Steps to reproduce the behavior

There are two *different* routes that produce *different* crashes:

* Open the Muon Data Analysis interface
* Go to the home tab and select instrument MUSR
* Ensure that you have access to the archive
* Load run 62260
* In the Data Analysis tab “TF Asymmetry Mode” should be ticked
* Set the “Groups/Pairs to fit” to “Custom”
* A pop-up should appear and make sure that only “fwd” is ticked
* Close the pop-up
* Add a “Flat Background” and “GausOsc” functions.
* Run a fit

----

* Open the Muon Data Analysis interface
* Go to the home tab and select instrument MUSR
* Ensure that you have access to the archive
* Load run 62260
* In the Data Analysis tab “TF Asymmetry Mode” should be ticked
* Set the “Groups/Pairs to fit” to “Custom”
* A pop-up should appear and make sure that only “fwd” is ticked
* Close the pop-up
* “GausOsc” function only.
* Run a fit

### Platforms affected

OSX, Windows

### PR DESCRIPTION
**Pull request by AnthonyLim23**
Fixes the bugs found in Muon Analysis due to TF asymmetry mode. 

**To test:**

Muon Analysis with 1 workspace in multi fit
1. Open muon analysis
2. Make sure multiple fitting is on in the settings tab
3. On the home tab load MUSR 62260
4. Go to data analysis
5. Change the group selection to custom -> make sure only one is selected in the pop up (e.g. fwd)
6. Add a fitting function
7. Turn on TFAsymm mode
8. It should now work (not crash)

No function in Muon Analysis
1. Open muon analysis
2. Make sure multiple fitting is off in the settings tab
3. On the home tab load MUSR 62260
4. Go to data analysis
5. Turn on TFAsymm mode
6. It should now work (not crash)
7. Clicking somewhere else will update the tick box

Zeros in calculateMuonAsymmetry
1. Open muon analysis
2. Make sure multiple fitting is off in the settings tab
3. On the home tab load MUSR 62260
4. Go to data analysis
5. Add a fitting function
6. Do a few fits to get good parameters
7. Turn on TFAsymm mode
8. Set the first flat background to 0 and fix it
9. Do a fit
10. It should have a useful error message at the top of the fitting menu

Fixes #23031. 

Does this update require release notes?
- [ ] Yes
- [ x] No

The release notes had some bits missing, so I have also updated them (not tested the links)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by NickDraper** [comment]
Crash still present on following the first testing example

MantidQtWidgetsCommonQt4.dll!MantidQt::MantidWidgets::MuonFitPropertyBrowser::ConvertFitFunctionForMuonTFAsymmetry(bool enabled) Line 1194

**Review by NickDraper — DISMISSED** [review]
Crash on 1st example

**Comment by AnthonyLim23** [comment]
is the crash the first of my examples or from the issue?

**Comment by NickDraper** [comment]
Anthony Ignore that crash comment, the brach had not merged cleanly so I was testing the previous state.

File: qt/widgets/common/src/MuonFitPropertyBrowser.cpp [hunk 1]
```diff
@@ -1243,18 +1248,20 @@ void MuonFitPropertyBrowser::ConvertFitFunctionForMuonTFAsymmetry(
...
+  if (old->nParams() > 0) {
+    ConvertFitFunctionForMuonTFAsymmetry(enabled);
+    // Show or hide the TFAsymmetry fit
+    if (enabled) {
+      m_settingsGroup->property()->addSubProperty(m_keepNorm);
+    } else {
+      m_settingsGroup->property()->removeSubProperty(m_keepNorm);
+    }
+  } else if (enabled) {
+    g_log.warning("No fitting function provided. Please add a function");
```
Comment (NickDraper) [inline qt/widgets/common/src/MuonFitPropertyBrowser.cpp]: This is too invisible, and does not mention what it is in relation to, or that TF asymmetry mode has been turned off.

I suggest, you use a messagebox with a message like
"No fitting function provided. ~TF Asymmetry mode needs a function added before enabling, Add a function and enable TF Asymmetry Mode again."

**Review by NickDraper — CHANGES_REQUESTED** [review]

**Comment by NickDraper** [comment]
Another crash:

1. Run example 1
1. Go to the settings tab
1. Disable multiple fitting
1. Return to the data analysis tab
1. Bang.  	Null pointer exception MantidMuon.dll!Mantid::Muon::ConvertFitFunctionForMuonTFAsymmetry::extractUserFunction(const boost::shared_ptr<Mantid::API::IFunction> & TFFuncIn) Line 254

**Comment by NickDraper** [comment]
Example 3 - Is this the error message you were expecting?
image

Deviations from testing notes:

1. It isn't at the top of the fitting menu
1. Calling this a useful error message is quite a stretch

**Review by NickDraper — APPROVED** [review]

**Comment by NickDraper** [comment]
Good to go, some areas for future improvement, but this is good enough for the release

**Review by martyngigg — CHANGES_REQUESTED** [review]
There is a documentation build warning in the muon.rst

**Review by martyngigg — APPROVED** [review]

**Pull request merged**